src/inmanta/server/services/compilerservice.py

```diff
@@ -601,13 +602,104 @@ async def request_recompile(
         Recompile an environment in a different thread and taking wait time into account.
 
         :param notify_failed_compile: if set to True, errors during compilation will be notified using the
-        "failed_compile_message".
-        if set to false, nothing will be notified. If not set then the default notifications are
-        sent (failed pull stage and errors during the do_export)
+                                      "failed_compile_message".
+                                      if set to false, nothing will be notified. If not set then the default notifications are
+                                      sent (failed pull stage and errors during the do_export)
         :param failed_compile_message: the message used in notifications if notify_failed_compile is set to True.
         :return: the compile id of the requested compile and any warnings produced during the request
+        """
+        return await self._request_recompile(
+            env,
+            force_update,
+            do_export,
+            remote_id,
+            metadata,
+            env_vars,
+            partial,
+            removed_resource_sets,
+            exporter_plugin,
+            notify_failed_compile,
+            failed_compile_message,
+        )
+
+    async def request_recompile_in_transaction(
+        self,
+        connection: Connection,
+        env: data.Environment,
+        force_update: bool,
+        do_export: bool,
+        remote_id: uuid.UUID,
+        metadata: Optional[JsonType] = None,
+        env_vars: Optional[Mapping[str, str]] = None,
+        partial: bool = False,
+        removed_resource_sets: Optional[List[str]] = None,
+        exporter_plugin: Optional[str] = None,
+        notify_failed_compile: Optional[bool] = None,
+        failed_compile_message: Optional[str] = None,
+    ) -> Tuple[Optional[uuid.UUID], Warnings]:
+        """
+        This method is similar to the `request_recompile()` method, but should be used when the compile request
+        is part of an ongoing database transaction. The given `connection` attribute should be within the scope
+        of that database transaction. It is required to call `CompileService.notify_compile_request_committed()`
+        right after the transaction commits.
+        """
+        return await self._request_recompile(
+            env,
+            force_update,
+            do_export,
+            remote_id,
+            metadata,
+            env_vars,
+            partial,
+            removed_resource_sets,
+            exporter_plugin,
+            notify_failed_compile,
+            failed_compile_message,
+            connection=connection,
+            in_db_transaction=True,
+        )
+
+    async def notify_compile_request_committed(self, compile_id: uuid.UUID) -> None:
```

Comment:
I have no strict objection but I think we could get rid of the `compile_id` parameter if we wanted to. All that needs to be done here is to check whether there's a compile running and if not whether there's a compile queued. It's very similar to `_dequeue()`, the only difference being that that method has as a precondition that no compile is running but we could easily extract the common part if we wanted to.

Reply:
Then you want to replace `compile_id` by `environment_id`. Or would you execute that logic for every environment?

Reply:
`environment_id` indeed.